[Embeddable Refactor] Create Decoupled Presentation Panel

ThomThomson opened the pull request:
Closes https://github.com/elastic/kibana/issues/167426
Closes https://github.com/elastic/kibana/issues/167890

# Summary
Part of https://github.com/elastic/kibana/issues/167429

This PR contains much of the prep work required to decouple the Embeddables system from Kibana in anticipation of its deprecation and removal.

To accomplish this, several new systems and concepts have been added to Kibana which will eventually replace portions of the Embeddable framework. An overview of each new concept can be found in this PR description. If you would like more information, the [architecture document](https://docs.google.com/document/d/1pD5XiCtL36S0tnkEE04xDbyfToh80KNp9WaKdzSXNeU/edit) goes into more detail.

# The Presentation Panel
Previously, the generic _frame_ component that linked React components to the UI Actions library was tightly coupled to the Embeddables system and the Embeddables plugin. This PR removes that coupling.

<img width="624" alt="Screenshot 2024-01-10 at 3 24 18 PM" src="https://github.com/elastic/kibana/assets/14276393/3873930a-9c68-41e5-a46e-239deca46acb">

Now, any React component can be linked to the UI Actions library by passing it as a prop to the newly created `PresentationPanel` compoent. This linkage is created via an API object, which is passed from the component via the [use imperative handle](https://react.dev/reference/react/useImperativeHandle) hook. Methods and state on this API will be used by the Presentation Panel. For instance, if the API `publishes` a title, the latest value of that title will be displayed in the panel header.

# Publishing Subjects
This API system works great as-is for completely imperative things like functions, `onEdit` for instance, but for state which can change, we need a standardized system which can publish individual pieces of state in a reactive form. To fulfil this requirement, this PR introduces Publishing subjects.

Publishing subjects are essentially _read only_ RXJS behavior subjects. This allows APIs to pass around changeable pieces of state, react to those changes, and get the last published value.

> [!NOTE]  
> Publishing subjects are meant to be readonly because they are **not** meant to be a source of truth. Instead, they are meant to forward or publish state from _another_ source of truth, whether it's from a Use State, some other RXJS subject, from Redux, or some other state management system.

A number of convenience functions are included to allow translating state from React components, or state from legacy Embeddables into Publishing Subjects. Additionally, hooks are included to allow components to render the latest state from the publishing subjects.

# UI Actions changes
A number of changes have been made to both the UI Actions framework itself, and a number of Actions registered with those frameworks.
 
### Dashboard & Presentation Panel Actions
Because UI Actions can now theoretically be linked to _any component_ via the new Presentation Panel component, actions which were tightly coupled with the old Embeddable system should be refactored to list their actual dependencies declaratively, which effectively decouples them from the embeddable system.

For instance, instead of the `edit panel action` requiring an `Embeddable` instance as its context, it instead requires any API object with an `onEdit` method and a `viewMode` currently set to `edit`. This allows for more clarity in requirements, and allows for more flexibility in what each action actually _does_. 

> [!NOTE]  
> It's important to note that this PR leaves some code in place to ensure interoperability between Actions which are still tightly coupled to Embeddables, and actions which have been updated. This includes:
>  1. The name of the key in all context objects is still `embeddable`. This should eventually be updated to `api`
>  2. The new API methods and state accessors are simply appended onto the existing Embeddable instances. This means that actions which rely on legacy methods can do so alongside actions that rely on the new methods.

### Subscribe to compatibility changes
Two new optional methods have been added to the UI Action definition: `subscribeToCompatibilityChanges` and `couldBecomeCompatible`. These methods are meant to allow actions to declare exactly which pieces of state from a given API will cause them to become compatible or incompatible.

Previously, _any_ state change to _any_ embeddable or its parent would cause the entire tree of Actions to re-calculate their compatibility state. With these changes in place, that code has been removed, and any action which is expected to frequently change its compatibility state in response to some other state can do so via this system.

# Architecture / Progress
This PR accomplishes some of the items on our roadmap towards the complicated architecture which will temporarily exist to ease our transition towards the new Embeddables framework. Below, you can see the pieces of this architecture added by this PR, and the pieces which will be added by the followup.

![unnamed](https://github.com/elastic/kibana/assets/14276393/b309c0c3-6ae8-4fd9-b9ef-ee562be99d06)

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

ThomThomson commented:
/ci

nreese commented on src/plugins/dashboard/public/dashboard_actions/add_to_library_action.tsx:
```diff
@@ -6,110 +6,70 @@
+        'data-test-subj': 'addPanelToLibrarySuccess',
+      });
+    } catch (e) {
+      this.toastsService.addDanger({
```
Should error message also include panel title?

ThomThomson replied:
Good call, done in [b9f6ca7](https://github.com/elastic/kibana/pull/172017/commits/b9f6ca76d7c77d708667a7b2ec643d1ddbbb07db)

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

Ikuni17 reviewed (approved):
`limits.yml` only

nreese commented:
filters notification is still getting displayed when in view mode

<img width="500" alt="Screenshot 2024-01-11 at 4 18 42 PM" src="https://github.com/elastic/kibana/assets/373691/41d3f4bd-1169-4927-a928-57914ea0b78c">

Steps to problem
1. create new dashboard
2. add maps panel with one documents layer and filter. Click "Save and return"
3. save dashboard
4. switch to view mode. Filter notification should not be showing

Something else weird, if you then reload the dashboard, the filter notification is gone and if you switch to edit mode, the filter notification is not displayed.

rshen91 reviewed (approved):
Tested dashboard locally stateful and stateless - this work hasn't broken any functionality. Great additions!

TinaHeiligers reviewed (approved):
`.jsonc` file changes LGTM

ThomThomson commented:
@nreese 
> filters notification is still getting displayed when in view mode

Good catch! Fixed in [2d0162b](https://github.com/elastic/kibana/pull/172017/commits/2d0162babe2ed101f00dbe80aaf0d717b82fd649)

The problem was:

1. RXJS merge took in an array and for some reason the types system didn't catch this as a problem. I'm not able to figure out why this syntax was allowed, but here's a [thread](https://github.com/ReactiveX/rxjs/issues/1308) discussing it.
2. The test `src/plugins/dashboard/public/dashboard_actions/filters_notification_action.test.tsx` missed this because some of the lines were the wrong way around. In order to properly test the expected behaviour, it now subscribes _before_ changing the value and checking that the change is picked up.

davismcphee reviewed (approved):
Code-only review, saved search embeddable changes LGTM 👍

angorayc reviewed (approved):
Thanks for updating this. Tested locally, LGTM.

mbondyra commented on x-pack/plugins/lens/public/embeddable/embeddable.tsx:
```diff
@@ -1452,7 +1452,7 @@ export class Embeddable
     ).forEach((dataView) => indexPatterns.push(dataView));
 
-    this.dataViews = uniqBy(indexPatterns, 'id');
+    this.internalDataViews = uniqBy(indexPatterns, 'id');
```
what is the difference between `this.dataViews.value` and `this.internalDataViews` that we have to load here? Maybe we could just use the embeddable already existing `this.dataViews.value`?

ThomThomson replied:
I can look into this in more detail, but `this.dataViews.value` is the same as calling `this.getOutput().indexPatterns`. I wasn't exactly sure why Lens had a separate `dataViews` key instead of using index patterns from the output, so I've left it to keep the behaviour the exact same.

I'm planning on cleaning this all up during the conversion of Lens to actually _use_ the new Embeddable system.

mbondyra reviewed (approved):
Left a comment, but I am approving this PR anyway because maybe it's better to apply a change I mentioned separately. Good job!

dej611 commented on src/plugins/dashboard/public/dashboard_actions/replace_panel_flyout.tsx:
```diff
@@ -92,7 +77,9 @@ export class ReplacePanelFlyout extends React.Component<Props> {
 
-    const panelToReplace = 'Replace panel ' + this.props.panelToRemove.getTitle() + ' with:';
+    const panelToReplace =
+      'Replace panel ' + this.props.api.panelTitle?.value ??
```
Just noticed this string is not passing thru i18n.

ThomThomson replied:
Good catch, fixed this!

dej611 reviewed (commented):
Minor changes (probably not blocking)

#### Inspect panel title

I found that for some panels, if they are saved with no title + show title option, unless the panel settings is edited, then inspector will show a `untitled` label in the inspector:

![dashboard_inspect_title_bug](https://github.com/elastic/kibana/assets/924948/0585ff43-2f05-4381-8834-a959c26066a8)

I've tested in `main` and `untitled` never shows up, no matter the title setting:
![dashboard_inspect_title](https://github.com/elastic/kibana/assets/924948/b132d6be-bb9e-455f-a3fa-3eeac4e7e2d8)

This happens only on untouched panels. Not sure the behaviour in `main` is correct (to never show `untitled` as I see there are pros and cons with it), but this one here is inconsistent.

#### Lens playground embeddable options

![lens_playground_bug](https://github.com/elastic/kibana/assets/924948/1d029036-5622-4e91-b9a3-e0c5339dc300)

when enabling the default actions in the embeddable panel and forcing a re-render I see this error in console:
```
Uncaught (in promise) TypeError: Cannot use 'in' operator to search for 'getVis' in undefined
    at isVisualizeEmbeddable (edit_in_lens_action.tsx:35:1)
    at EditInLensAction.isCompatible (edit_in_lens_action.tsx:104:1)
    at ActionInternal.isCompatible (action_internal.ts:56:1)
    at ui_actions_service.ts:96:1
    at Array.map (<anonymous>)
    at UiActionsService.getTriggerCompatibleActions (ui_actions_service.ts:96:1)
    at getActions (embeddable_component.tsx:90:1)
    at presentation_panel_context_menu.tsx:56:1
    at presentation_panel_context_menu.tsx:62:1
    at presentation_panel_context_menu.tsx:96:1
```

#### Saved visualization icon

Where's gone the saved visualization icon?

<img width="1507" alt="Screenshot 2024-01-16 at 11 19 35" src="https://github.com/elastic/kibana/assets/924948/614302af-893d-442c-b185-7d1f4fe289ce">

Is it part of a redesign?
I can see it in `main` being shown:

<img width="1494" alt="Screenshot 2024-01-16 at 11 25 01" src="https://github.com/elastic/kibana/assets/924948/f28ed0d6-a912-4a68-b5b6-3616286b73d3">

nreese commented on src/plugins/dashboard/public/dashboard_actions/unlink_from_library_action.tsx:
```diff
@@ -41,64 +42,37 @@ export class UnlinkFromLibraryAction implements Action<UnlinkFromLibraryActionCo
+        'data-test-subj': 'unlinkPanelSuccess',
+      });
+    } catch (e) {
+      this.toastsService.addDanger({
```
How about including title in failure message as well

ThomThomson replied:
Done

ThomThomson commented on packages/presentation/presentation_publishing/interfaces/publishes_local_unified_search.ts:
```diff
@@ -0,0 +1,77 @@
+): unknownApi is PublishesLocalUnifiedSearch => {
+  return Boolean(
+    unknownApi &&
+      (unknownApi as PublishesLocalUnifiedSearch)?.localTimeRange !== undefined &&
```
These are really useful generic functions for reducing the amount of code in these package! I've been going back and forth on whether or not to use functions like this for type checking.

I think there is a tradeoff here:
- This solution is less code, but more difficult to grok at first glance.
- The current solution in the PR is more verbose, but might be easier to read.

See a very similar discussion with @davismcphee in the architecture document [here](https://docs.google.com/document/d/1pD5XiCtL36S0tnkEE04xDbyfToh80KNp9WaKdzSXNeU/edit?disco=AAAA-q3V_9c)

ThomThomson commented:
@dej611 thank you for the very thorough review here! 

## Untitled in inspector
Looks like I updated the logic in the `open_inspector` action a little here which is causing this. Basically before, the flyout title and the filename were different, with the flyout title only using the Embeddable's `getTitle`, but the filename falling back to `untitled` eventually. I updated that so that both of them fall back to `untitled` when nothing is available.

Before
``` typescript
const session = this.inspector.open(adapters, {
      title: embeddable.getTitle(),
      options: {
        fileName:
          embeddable.getTitle() || // pick the visible title
          embeddable.getInput().title || // or the custom title if used, but currently hidden
          embeddable.getOutput().defaultTitle || // or the saved title
          // in the very last resort use "untitled"
          i18n.translate('embeddableApi.panel.inspectPanel.untitledEmbeddableFilename', {
            defaultMessage: 'untitled',
          }),
      },
    });
```

After
``` typescript
const panelTitle =
      embeddable.panelTitle?.value ??
      i18n.translate('presentationPanel.action.inspectPanel.untitledEmbeddableFilename', {
        defaultMessage: 'untitled',
      });
const session = inspector.open(adapters, {
      title: panelTitle,
      options: {
        fileName: panelTitle,
      },
    });
```

I can revert this change if you'd like.

## Lens playground errors when enabling default actions
Great catch! I didn't think to check inside the Lens playground for regressions, but it turned out I was sending the wrong key to the UI Actions when a custom actions callback was provided.

## Missing saved visualization icon
I can't reproduce this. The icon should appear in edit mode and should disappear in view mode. This gif is recorded on the feature branch: 

![Icon](https://github.com/elastic/kibana/assets/14276393/31008c21-ac3b-4fff-ba7c-e78fdb3e0111)

Let me see how this worked on Main - maybe there was some issue which made this show up in view mode also?

ThomThomson commented on src/plugins/dashboard/public/dashboard_actions/library_notification_action.test.tsx:
```diff
@@ -6,91 +6,57 @@
-  embeddable.updateInput(await embeddable.getInputAsRefType());
-  embeddable.updateInput({ viewMode: ViewMode.VIEW });
-  expect(await action.isCompatible({ embeddable })).toBe(false);
+  it('is incompatible when api publishes a saved object id', async () => {});
```
Good call out - I've removed this test because it is a leftover from when I was checking specifically for saved object IDs, rather than letting the API itself determine what `unlink` means.

This is covered by the test above `'is incompatible when can unlink from library resolves to false'`

ThomThomson commented on src/plugins/dashboard/public/dashboard_actions/library_notification_action.test.tsx:
```diff
@@ -6,91 +6,57 @@
-});
+  it('calls onChange when view mode changes', async () => {
+    const onChange = jest.fn();
+    updateViewMode('view');
```
Good catch, done!

ThomThomson commented on src/plugins/dashboard/public/dashboard_actions/filters_notification_popover.tsx:
```diff
@@ -6,50 +6,66 @@
+        queryLanguage: language,
+      };
+    }
+  }, [api, setDisableEditButton]);
```
Great question, we'll definitely need to refactor this once local queries / filters are editable inline.

But for the time being, we can be certain that neither the API nor the `localQuery` will change reference while this popover is open. Really, the `api` is only there to satisfy the typescript rule exhaustive deps to avoid using an `// @ts-ignore`.

If we anticipated that the query would change over the lifetime of the flyout, we'd likely need to call `useLocalQuery` from `packages/presentation/presentation_publishing/interfaces/publishes_local_unified_search.ts` and then extract the string and language from there like:

```typescript
  const localQuery = useLocalQuery(api);
  const { queryString, queryLanguage } = useMemo(() => {
    if (!localQuery) return {};
    if (isOfQueryType(localQuery)) {
      if (typeof localQuery.query === 'string') {
        return { queryString: localQuery.query };
      } else {
        return { queryString: JSON.stringify(localQuery.query, null, 2) };
      }
    } else {
      setDisableEditButton(true);
      const language: 'sql' | 'esql' | undefined = getAggregateQueryMode(localQuery);
      return {
        queryString: localQuery[language as keyof AggregateQuery],
        queryLanguage: language,
      };
    }
  }, [localQuery, setDisableEditButton]);
```

Currently, this feels like overkill, because it would introduce an unnecessary subscription.

ThomThomson commented on src/plugins/presentation_panel/public/panel_component/presentation_panel_error.tsx:
```diff
@@ -0,0 +1,103 @@
+      iconType="warning"
+      iconColor="danger"
+      layout="vertical"
+      actions={
```
Good catch, definitely a bad merge. I've fixed this

nreese commented on src/plugins/embeddable/public/lib/embeddables/compatibility/edit_legacy_embeddable.tsx:
```diff
@@ -0,0 +1,111 @@
+import { isExplicitInputWithAttributes } from '../embeddable_factory';
+import { EmbeddableInput } from '../i_embeddable';
+
+const getLatestAppId = async (): Promise<string | undefined> => {
```
Could this be simplified to just `return core.application.currentAppId$.pipe(first()).toPromise();`

nreese replied:
Actually, toPromise is deprecated. instead, you could just use `await lastValueFrom(core.application.currentAppId$)`

ThomThomson replied:
I'm not sure that `lastValueFrom` will work properly in this case. I think it waits for the observable to complete.

ThomThomson replied:
Good call, `firstValueFrom` works perfectly here and makes the code much shorter!

nreese commented on src/plugins/embeddable/public/lib/embeddables/embeddable.tsx:
```diff
@@ -253,21 +337,34 @@ export abstract class Embeddable<
   }
 
+  public async untilInitializationFinished(): Promise<void> {
+    return new Promise((resolve) => {
```
How about just `return lastValueFrom
 (this.initializationFinished)`

nreese replied:
How about using firstValueFrom then?

ThomThomson replied:
can confirm that `lastValueFrom` worked great here! Thanks for the suggestion!

ThomThomson replied:
I had to revert this change because of a jest test failure. I've opened a followup [issue](https://github.com/elastic/kibana/issues/175068) to see what went wrong here.

ThomThomson commented on src/plugins/inspector/public/adapters/has_inspector_adapters.ts:
```diff
@@ -0,0 +1,24 @@
+/**
+ * The interface that the adapters used to open an inspector have to fullfill.
+ */
+export interface Adapters {
```
Good call, updated that! Best not to duplicate the `any` here.

nreese commented on src/plugins/inspector/public/index.ts:
```diff
@@ -18,7 +18,12 @@ export function plugin(initializerContext: PluginInitializerContext) {
 export { InspectorPublicPlugin as Plugin } from './plugin';
+export type { Setup, Start } from './plugin';
 export * from './types';
-export * from '../common/adapters';
```
you will need to export Adapters from common/adapters

nreese commented on src/plugins/presentation_panel/public/panel_actions/customize_panel_action/custom_time_range_badge.test.ts:
```diff
@@ -0,0 +1,61 @@
+    const onChange = jest.fn();
+    updateTimeRange(mockTimeRange);
+    updateTimeRange(undefined);
+    action.subscribeToCompatibilityChanges(context, onChange);
```
Move updates to after subscriptions is created

ThomThomson replied:
Done!

nreese commented on src/plugins/presentation_panel/public/panel_actions/edit_panel_action/edit_panel_action.test.tsx:
```diff
@@ -0,0 +1,71 @@
+  it('calls onChange when view mode changes', () => {
+    const onChange = jest.fn();
+    updateViewMode('view');
+    action.subscribeToCompatibilityChanges(context, onChange);
```
Call updateViewMode after creating subscription

ThomThomson replied:
Done!

nreese commented on src/plugins/presentation_panel/public/panel_component/presentation_panel.tsx:
```diff
@@ -0,0 +1,61 @@
+    );
+  }
+
+  if (loading || !value?.Panel || !value?.unwrappedComponent)
```
Should an error state be displayed if loading=false and value.Panel or value.unwrappedComponent do not exist? In this case, the panel will get stuck showing loading screen forever. I think an error would help, especially for developers, who are the ones most likely to trigger this case.

ThomThomson replied:
Good call. I've updated the error section above to also trigger when `(!loading && (!value?.Panel || !value?.unwrappedComponent))`

nreese commented on src/plugins/presentation_panel/public/panel_component/presentation_panel_internal.tsx:
```diff
@@ -0,0 +1,138 @@
+
+  const [initialLoadComplete, setInitialLoadComplete] = useState(!dataLoading);
+  if (dataLoading === false && !initialLoadComplete) {
+    setInitialLoadComplete(true);
```
Should setInitialLoadComplete be inside an effect. State should not be set from rendering.

ThomThomson replied:
This code was just copied over from https://github.com/elastic/kibana/pull/171238. Refer to [this comment](https://github.com/elastic/kibana/pull/171238/files#r1403113178)  for more info on why it was done this way. I'd be okay with wrapping this in a `useEffect`.

nreese reviewed (approved):
kibana-presentation changes LGTM other than some minor comments. This is an impressive accomplishment and is a huge step forward for the new embeddable system. Thanks for all of the hard work and high quality changes.

code review, tested in chrome
